src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -96,3 +116,153 @@ void ShenandoahGenerationalHeap::stop() {
   regulator_thread()->stop();
   ShenandoahHeap::stop();
 }
+
+ShenandoahGenerationalHeap::TransferResult ShenandoahGenerationalHeap::balance_generations() {
+  shenandoah_assert_heaplocked_or_safepoint();
+
+  ShenandoahOldGeneration* old_gen = old_generation();
+  size_t old_region_surplus = old_gen->get_region_surplus();
+  size_t old_region_deficit = old_gen->get_region_deficit();
+  old_gen->set_region_surplus(0);
+  old_gen->set_region_deficit(0);
+
+  if (old_region_surplus) {
+    bool success = generation_sizer()->transfer_to_young(old_region_surplus);
+    return TransferResult {
+      success, old_region_surplus, "young"
+    };
+  }
+
+  if (old_region_deficit) {
+    bool success = generation_sizer()->transfer_to_old(old_region_deficit);
+    if (!success) {
+      old_gen->handle_failed_transfer();
+    }
+    return TransferResult {
+      success, old_region_deficit, "old"
+    };
+  }
```

Comment:
Here again, a single field with sign and a sign comparison would work fine and may be even make it more natural.